zerver/actions/message_flags.py

```diff
@@ -80,30 +83,31 @@ def do_mark_stream_messages_as_read(
 ) -> int:
     log_statsd_event("mark_stream_as_read")
 
-    msgs = UserMessage.objects.filter(
-        user_profile=user_profile,
-    )
+    with transaction.atomic(savepoint=False):
+        query = UserMessage.select_for_update_query().filter(
+            user_profile=user_profile,
+        )
 
-    msgs = msgs.filter(message__recipient_id=stream_recipient_id)
+        query = query.filter(message__recipient_id=stream_recipient_id)
 
-    if topic_name:
-        msgs = filter_by_topic_name_via_message(
-            query=msgs,
-            topic_name=topic_name,
-        )
+        if topic_name:
+            query = filter_by_topic_name_via_message(
+                query=query,
+                topic_name=topic_name,
+            )
 
-    msgs = msgs.extra(
-        where=[UserMessage.where_unread()],
-    )
+        query = query.extra(
+            where=[UserMessage.where_unread()],
+        )
```

Comment:
I'm going to fix this when pushing, but it'd be cleaner to inline these extra unconditional filters into the original `query` definition.